### PR TITLE
feat(gitGraph): Add interactive click support for commits, branches, and tags

### LINKED ISSUE #5599: Links in GitGraph ?
### Proposal

Hello
I'm looking to use links or buttons in GitGhrap, especially in commit ids.
Is this possible?
Thanks

### Example

gitGraph TB:
        commit id:"<a href='GitHub.com'>GitHub</a>"

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by mkobit**
## :bookmark_tabs: Summary

Adds interactive click functionality to `gitGraph` diagrams, enabling clickable links on commits, branches, and tags. Users can navigate to external URLs by clicking diagram elements, with support for custom tooltips and link targets.

Resolves #5599 

### 🫙 Example

The diagram below shows the structure.

````
```mermaid
gitGraph
  commit id: "no-click"
  commit id: "commit-same-tab"
  commit id: "commit-new-tab" tag: "tag-new-tab"
  branch branch-same-tab
  checkout branch-same-tab
  commit id: "normal"
  checkout main
  branch branch-new-tab
  checkout branch-new-tab
  commit id: "another" tag: "tag-same-tab"

  %% Click statements add interactivity (syntax shown below):
  click commit "commit-same-tab" "https://example.com/commit1" "same tab"
  click commit "commit-new-tab" "https://example.com/commit2" "new tab" _blank
  click branch "branch-same-tab" "https://example.com/branch1" "same tab"
  click branch "branch-new-tab" "https://example.com/branch2" "new tab" _blank
  click tag "tag-same-tab" "https://example.com/tag1" "same tab"
  click tag "tag-new-tab" "https://example.com/tag2" "new tab" _blank
```
````

**Syntax:** `click <type> <id> <url> ["tooltip"] [target]`

- `<type>`: `commit`, `branch`, or `tag`
- `<id>`: The element's identifier
- `<url>`: Target URL
- `["tooltip"]`: Optional hover text
- `[target]`: Optional link target (`_blank`, `_self`, `_parent`, `_top`)

### Following flowchart interaction patterns

The implementation mirrors the click interaction syntax from flowchart diagrams to maintain consistency across Mermaid diagram types:

- **Same syntax structure**: `click <identifier> <url> ["tooltip"] [target]`
- **Same link targets**: Supports `_self`, `_blank`, `_parent`, and `_top`
- **Same security model**: Respects `securityLevel` configuration (disabled in `strict`, enabled in `loose`)

### Type-specific element targeting

Unlike `flowchart` where node IDs are globally unique, gitGraph allows the same identifier to exist across different element types (a branch and commit can share an ID). The syntax requires explicit type specification to prevent ambiguity:

```
click commit "feature-123" "https://example.com"
click branch "feature-123" "https://example.com"
```

- Prevents targeting conflicts when IDs overlap
- Makes click statements self-documenting (clear which element is being linked)
- Aligns with `gitGraph`'s grammar where elements are type-qualified

### Security and accessibility

External links include security and accessibility attributes:
- **Security**: Links to different origins include `rel="noopener noreferrer"` to prevent tab-nabbing attacks
- **Accessibility**: Clickable elements include:
  - `role="link"` attribute for screen readers
  - `<title>` elements with tooltip text for hover feedback
  - Visual cursor changes to indicate interactivity

### Scope limitations

**JavaScript callbacks not implemented**: Unlike flowchart, this implementation only supports URL links, not JavaScript callback functions. This was a scope decision to deliver the core feature requested in #5599 (linking commits to external resources). Callback support can be added in future iterations if needed.

### Use case context (from #5599)

As described in the original feature request, clickable commits make _"Mermaid diagrams a powerful bridge between documentation and actual code review context."_ Teams can:
- Link commits to pull requests or code review tools
- Connect branches to CI/CD pipeline dashboards
- Link tags to release notes or deployment documentation
- Embed interactive git workflows in project documentation

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://mermaid.js.org/community/contributing.html)
- [x] :computer: have added necessary unit/e2e tests.
- [x] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://mermaid.js.org/community/contributing.html#update-documentation) is used for all new features.
- [x] :butterfly: If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

## Summary

Adds interactive click support to `gitGraph` diagrams for commits, branches, and tags.

### Highlights

- Adds `click <type> <id> <url> ["tooltip"] [target]` syntax.
- Supports `_blank`, `_self`, `_parent`, and `_top` targets.
- Handles overlapping identifiers by element type.
- Applies `securityLevel` rules:
  - `strict` disables links.
  - `sandbox` forces `_top`.
  - `antiscript` sanitizes unsafe URLs.
- Adds URL and tooltip sanitization.
- Adds `noopener noreferrer` protection for cross-origin links.
- Adds accessibility attributes, tooltips, focus styles, and clickable cursors.
- Updates the parser, renderer, documentation, demos, and changeset.
- Adds parser, database, renderer, integration, Playwright, and visual regression coverage.
- Addresses issue `#5599` by enabling links for commit IDs, branches, and tags.

Automatic tag-link generation and JavaScript callbacks remain out of scope.

### CODE REVIEW

**Comment by selescop** [comment]
Nice feature to have!

If I may add my proposal, would it be possible to have an automatic link on commits that have a 'tag' field to the actual tag <orga_name>/<repo_name>/releases/tag/<tag_name>?
For example:
commit id: "commit-new-tab" tag: "v10.9.5"
would automatically points to
https://github.com/mermaid-js/mermaid/releases/tag/v10.9.5

**Comment by mkobit** [comment]
> Nice feature to have!
> 
> If I may add my proposal, would it be possible to have an automatic link on commits that have a 'tag' field to the actual tag <orga_name>/<repo_name>/releases/tag/<tag_name>? For example: commit id: "commit-new-tab" tag: "v10.9.5" would automatically points to https://github.com/mermaid-js/mermaid/releases/tag/v10.9.5

I agree this is a good feature and something I thought about including. I decided to cut scope to first implement the overall functionality and syntax for link support. To me, the auto-linking functionality seems like additional options that would go into the [chart configuration options](https://mermaid.ai/open-source/syntax/gitgraph.html#gitgraph-specific-configuration-options) that specify something like a link pattern. For example, imagine something like:

```
---
tagLinkTemplate: https://github.com/mermaid-js/mermaid/releases/tag/{tag}
---
```

I haven't worked through the syntax and UX yet. For users with GHE or other things they want to link to, I would imagine it needs to be thought about a bit more with additional examples.

**Comment by mkobit** [comment]
looking to get some feedback here before i start endlessly dealing with merge conflicts

**Review by ashishjain0512 — COMMENTED** [review]
[sisyphus-bot]

## Review: feat(gitGraph): Add interactive click support for commits, branches, and tags

First, a sincere apology for the delay, @mkobit — this PR has been open for 92 days and that's far too long for a contribution of this quality. Thank you for your patience and persistence.

---

## What's working well

🎉 **[praise]** The security model is exemplary. Using native SVG `<a>` elements instead of JS `onclick` handlers, `sanitizeUrl` for URL validation, `sanitizeText` for tooltip sanitization, `CSS.escape()` for selector safety, and `rel="noopener"` for tab security — this is defense-in-depth done right. The explicit test cases for `javascript:` URL blocking and `<script>` tag stripping in tooltips show security was a first-class concern.

🎉 **[praise]** Outstanding test coverage — 3 new test files (`gitGraphAst.spec.ts`, `gitGraphLink.spec.ts`, `gitGraphRenderer.click.spec.ts`) covering DB operations, parser integration, and renderer behavior including edge cases (special characters in IDs, URL sanitization, link overwriting). Plus 9 E2E snapshot tests in `cypress/integration/rendering/gitGraph.spec.js`. This is exactly the level of thoroughness we want.

🎉 **[praise]** Clean grammar design in the Langium `.langium` file — the `click` syntax is intuitive (`click commit "id" "url" "tooltip" _blank`) and the target is properly constrained to valid HTML values (`_blank`, `_self`, `_parent`, `_top`). Using a dedicated statement type rather than allowing HTML in labels was the right architectural call.

🎉 **[praise]** Good use of data attributes (`data-commit-id`, `data-branch-name`, `data-tag-name`) for element selection rather than fragile class-based or positional selectors. Defensive copy pattern on `getLinks()` prevents external mutation of internal state.

🎉 **[praise]** Documentation and changeset both included with correct `minor` bump and clear examples showing all three click types (commit, branch, tag) with optional tooltip and target.

---

## Items to address

### 🟡 [important] No explicit `securityLevel` guard in `setupClickEvents()`

`gitGraphRenderer.ts` — The `securityLevel` parameter is accepted by `setupClickEvents()` but never checked. The feature relies on the implicit assumption that click statements won't produce links in strict mode, but there's no explicit guard. If someone adds click statements to a diagram rendered in `strict` mode, the links would still be bound.

**Suggested fix:** Add an explicit guard at the top of `setupClickEvents()`:
```typescript
if (securityLevel === 'strict') {
  log.debug('Click events disabled in strict securityLevel');
  return undefined;
}
```

This matches how flowchart handles it and makes the security boundary explicit rather than implicit.

### 🟡 [important] `sanitizeUrl` re-exported from `utils.ts` — consider import path

`utils.ts` now re-exports `sanitizeUrl` from `@braintree/sanitize-url`. This makes it available to all modules. However, the flowchart click handler and other diagrams import `sanitizeUrl` directly from `@braintree/sanitize-url` rather than through utils. For consistency, it would be better to either:
1. Import directly from `@braintree/sanitize-url` (matching existing pattern), or
2. Have all diagrams import from `utils.ts` (but that's a broader refactor, out of scope)

Using the direct import avoids coupling the gitGraph feature to a new export from the shared utils module.

### 🟢 [nit] Sandbox mode target override is silent

`gitGraphRenderer.ts` — When in sandbox mode, the user-specified target is silently overridden to `_top`. A `log.debug()` message would help developers understand why their target setting is being ignored.

### 🟢 [nit] `Langium grammar change requires regeneration check`

The `.langium` grammar was modified. Per project conventions, `pnpm --filter parser langium:generate` must be run and the generated files in `packages/parser/src/language/generated/` must be committed. CI should verify this, but worth confirming the generated files are included (I don't see them in the changed files list — they may need to be regenerated).

---

## Security

No XSS or injection issues identified. The implementation follows security best practices:

- **URL sanitization:** `@braintree/sanitize-url` blocks `javascript:`, `data:`, `vbscript:` schemes
- **Tooltip sanitization:** `sanitizeText()` strips HTML tags
- **SVG `<a>` elements:** Native browser link behavior, not inline JS handlers
- **CSS selector escaping:** `CSS.escape()` with fallback regex prevents selector injection
- **`rel="noopener"`:** Prevents window.opener exploitation
- **DOMPurify:** Final SVG output still passes through DOMPurify sanitization

The only gap is the missing explicit `securityLevel` check (see 🟡 above), which is a defense-in-depth concern rather than an exploitable vulnerability.

---

## Self-check

- [x] At least one 🎉 [praise] item exists (5)
- [x] No duplicate comments
- [x] Severity tally: 0 🔴 / 2 🟡 / 2 🟢 / 0 💡 / 5 🎉
- [x] Verdict: COMMENT (2 🟡, no blocking items)
- [x] Not a draft PR
- [x] Tone check: appreciative, acknowledges delay, actionable

---

This is one of the best-implemented feature PRs I've reviewed — thorough security model, comprehensive tests, clean grammar design, and complete documentation. The two 🟡 items are straightforward. Let's get this across the finish line! 🚀

**Review by ashishjain0512 — COMMENTED** [review]
[sisyphus-bot]

## Supplementary Security Review: PR #7292

A deeper security pass surfaced a few additional findings worth addressing:

---

### 🟡 [important] `sanitizeUrl` return check is dead code

`gitGraphRenderer.ts` — The guard `if (!sanitizedUrl)` will **never trigger**. `@braintree/sanitize-url` returns `"about:blank"` for dangerous URLs, never `undefined`/`null`/`""`. The code works correctly (dangerous URLs get `about:blank` as href, which is benign), but the warning log is dead code.

**Suggested fix:** Check for the actual blocked value:
```typescript
const sanitizedUrl = sanitizeUrl(linkData.link);
if (sanitizedUrl === 'about:blank') {
  log.warn(`Blocked dangerous URL for ${linkData.type} "${id}": ${linkData.link}`);
  return; // Skip this link entirely rather than creating an about:blank anchor
}
```

Alternatively, consider using `utils.formatUrl()` which wraps `sanitizeUrl` and additionally respects `securityLevel` — this is what the flowchart diagram uses for its click handling.

### 🟡 [important] `clickable` CSS class never added — hover styles won't apply

`styles.js` defines extensive CSS rules for `.commit.clickable`, `.branchLabel.clickable`, and `.tag.clickable` (cursor, hover effects, focus styles). But `setupClickEvents()` in the renderer **never adds the `clickable` class** to elements that receive anchors. The flowchart equivalent explicitly calls `this.setClass(ids, 'clickable')`.

This means there's no visual indication that an element is clickable — reducing usability and accessibility. Users won't know they can click until they happen to mouse over.

**Suggested fix:** Add the `clickable` class to each element group when wrapping it in an `<a>`:
```typescript
element.classed('clickable', true);
```

### 🟢 [nit] Missing `noreferrer` on `rel` attribute

The anchor uses `rel="noopener"` but not `noreferrer`. Adding `rel="noopener noreferrer"` would also suppress the `Referer` header — a minor privacy improvement, especially for diagrams embedding links to external sites.

---

These don't change the overall verdict (COMMENT with 2 🟡 from the main review + 2 🟡 here = 4 🟡 total). The security model is fundamentally sound — `sanitizeUrl` blocks dangerous schemes, SVG `<a>` elements go through DOMPurify, and the architecture is correct. These are defense-in-depth improvements.

**Comment by mkobit** [comment]
Sorry it took me so long to get back to this!

**Comment by mkobit** [comment]
Demo of clickable tags, git commits of different types, and different link open types

**Comment by pbrolin47** [comment]
Hi @mkobit, thanks for addressing the issues from the review by @ashishjain0512 . 

[sisyphos-bot]

---
  What's Working Well
  
  🎉 [praise] Every single item from the two prior bot reviews has been addressed, cleanly. The securityLevel === 'strict' guard in addInteraction, the about:blank check, the
  clickable class, rel="noopener noreferrer", and the sandbox debug log are all correct. This shows real commitment to getting the implementation right.

  🎉 [praise] The patch 4 refactor — changing addInteraction to return undefined instead of a bare <g> when there's no interaction — is the right architectural call. The const
   parent = interactiveGroup || gBullets pattern keeps non-interactive elements structurally equivalent to the pre-PR behavior, which is exactly why Argos shows no
  regressions.
  
  🎉 [praise] The three-file test suite (gitGraphAst.spec.ts, gitGraphLink.spec.ts, gitGraphRenderer.click.spec.ts) is thorough: DB unit tests, full parser round-trips
  including all four target values, and renderer integration tests covering strict/loose/sandbox security levels plus the branch+tag case.

  ---
  Items to Address
  
  🟡 [important] ClickTarget grammar accepts arbitrary strings via STRING — enables frame hijacking

  packages/parser/src/language/gitGraph/gitGraph.langium — The grammar rule:

  ClickTarget returns string:
      STRING | '_blank' | '_self' | '_parent' | '_top';

  The STRING alternative makes the four keyword alternatives purely cosmetic — any quoted string is accepted. The value flows directly into a.attr('target', linkData.target)
  in addInteraction. While this isn't an XSS vector, a target attribute set to an arbitrary browsing context name (e.g., "attacker-window") enables tab-napping in environments
   where named frames exist — a real concern for a library embedded on GitHub/GitLab.
  
  Fix: Remove STRING from ClickTarget so only the four safe keywords are accepted:

  ClickTarget returns string:
      '_blank' | '_self' | '_parent' | '_top';

  🟡 [important] No runtime allowlist for target in addInteraction

  gitGraphRenderer.ts — There's no runtime check that linkData.target is one of the four safe values before it's set as an attribute. The as any casts in parseClick mean the
  TypeScript union type isn't enforced at runtime. If a target value other than the four safe values reaches addInteraction (e.g., via the grammar gap above, or future code
  paths), it gets set as-is.
  
  Fix: Add a runtime allowlist in addInteraction as defense-in-depth:

  const SAFE_TARGETS = new Set(['_self', '_blank', '_parent', '_top'] as const);
  // ...
  if (linkData.target && SAFE_TARGETS.has(linkData.target)) {
    a.attr('target', linkData.target);
  }
  
  This is independent of the grammar fix and matches the pattern in flowDb.

  🟡 [important] No Cypress E2E snapshot for a diagram with click statements

  The existing E2E tests pass (no regression), which is great. But per the project's test strategy, renderer/style changes require E2E visual tests — and there's no baseline
  snapshot showing a gitGraph diagram WITH click interactions rendering correctly. The styles.js additions (cursor, hover scale, transition) and the anchor-wrapping DOM
  changes are only covered by unit tests with mocked getBBox.
  
  A single imgSnapshotTest in cypress/integration/rendering/git/gitGraph.spec.js using the demo diagram from the PR description would provide a visual regression baseline for
  future changes:

  it('should render gitGraph with click interactions', () => {
    imgSnapshotTest(`
      gitGraph
        commit id: "ONE"
        click commit "ONE" "https://example.com" "Link"
        branch dev
        commit id: "TWO"
        click branch "dev" "https://example.com/dev"
    `, {});
  });

  ---
  Nits
  
  🟢 [nit] href is optional in grammar but required in docs/syntax

  The Langium grammar has (href=STRING)? making the URL optional, but the PR description and docs document it as required (click <type> <id> <url>). A user who writes click 
  commit "id" with no URL gets no error and no link — silent failure. Either enforce href=STRING (required, matching docs), or document the no-URL no-op behavior explicitly.

  🟢 [nit] as any casts in parseClick

  db.setLink?.(click.id, click.href ?? '', click.type as any, click.tooltip, click.target as any);

  Once the grammar is fixed (removing STRING from ClickTarget), the generated AST will type target as '_blank' | '_self' | '_parent' | '_top' and these casts should be
  removable. Worth revisiting after the grammar fix.

  🟢 [nit] sanitizeUrl vs formatUrl consistency

  addInteraction imports and uses sanitizeUrl directly from @braintree/sanitize-url, while the rest of the codebase (flowchart, class, state diagrams) uses utils.formatUrl()
  which wraps it. This is accidentally stricter (sanitizes even in loose mode), not a security regression. Consider switching to formatUrl for consistency and so the security
  contract is explicit.

  ---
  Security

  The core security model is sound:
  - sanitizeUrl blocks javascript:, data:, vbscript: and handles obfuscation via iterative decode
  - Tooltip goes through sanitizeText → DOMPurify then .text() (textContent, not innerHTML)
  - DOMPurify's XML allowlist passes xlink:href through but independently validates URI safety — javascript: can't survive both layers
  - rel="noopener noreferrer" is correctly set for cross-origin protection

  The two 🟡 security items above are defense-in-depth concerns (tab-napping via arbitrary frame names), not confirmed XSS vectors.

  ---
  Self-Check

  - At least one 🎉 praise item (3)
  - No duplicate comments
  - Severity tally: 0 🔴 / 3 🟡 / 3 🟢 / 0 💡 / 3 🎉
  - Verdict: REQUEST_CHANGES (3 🟡 issues)
  - Not a draft PR
  - Tone: community — warm, acknowledges good work, actionable

  ---
  This PR is genuinely close to landing. The implementation quality is high and the security thinking shows in the code. Three targeted fixes — the grammar constraint, a
  one-line Set.has() check, and a Cypress snapshot — should get it across the line. Let's do it! 🚀

  ---

File: cypress/integration/rendering/git/gitGraph.spec.js [hunk 1]
```diff
@@ -2179,4 +2179,21 @@ gitGraph TB:
...
+       commit id: "THREE"
+       checkout main
+       merge develop id: "FOUR"
+       click commit "ONE" "https://example.com/commit" "Commit Tooltip"
+       click branch "develop" "https://example.com/branch" "Branch Tooltip"
+       click tag "v1.0" "https://example.com/tag" "Tag Tooltip"
+      `,
+      { securityLevel: 'loose' }
+    );
+  });
```
Comment (mkobit) [inline cypress/integration/rendering/git/gitGraph.spec.js:2198]: i wasn't sure how to add a diff test that activates or handles the hover/clickable elements styling, so just added this test for now

**Comment by mkobit** [comment]
I think this is in a fairly good state, thanks for all the feedback! I know this is fairly large for an additive PR but hoping it can still get traction for inclusion!

**Comment by mkobit** [comment]
Looks like the ground moved from under this changeset's feet again.

I'm working on rebasing and resolving the conflicts.
